File: labs/Module_3_JAVA_LAB_2/Pin_Generator.java
import java.util.Scanner;

public class Pin_Generator {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        // System.out.println("Enter the numbers separated by spaces:");
        String input = scanner.nextLine();
        String[] numbers = input.split(" ");

        StringBuilder pin = new StringBuilder();
        for (String number : numbers) {
            int sum = 0;
            // Sum all digits in the number
            for (char digit : number.toCharArray()) {
                sum += digit - '0';
            }
            // Reduce to a single digit
            int singleDigit = reduceToSingleDigit(sum);
            // Convert to character or leave as digit based on even or odd
            char pinChar = convertToPINCharacter(singleDigit);
            pin.append(pinChar);
        }
        System.out.println( pin.toString());
        scanner.close();
    }

    // Helper method to reduce any number to a single digit
    private static int reduceToSingleDigit(int number) {
        while (number > 9) {
            int newSum = 0;
            while (number > 0) {
                newSum += number % 10;
                number /= 10;
            }
            number = newSum;
        }
        return number;
    }

    // Helper method to convert a single digit to its corresponding PIN character
    private static char convertToPINCharacter(int digit) {
        if (digit % 2 != 0) { // If it's odd
            // 'a' is 97 in ASCII, and '1' is mapped to 'a' which is 97 - 1
            return (char) ('a' + digit - 1);
        } else {
            return (char) ('0' + digit); // For even numbers, just return the digit itself
        }
    }
}
                                                                                                                            

File: labs/Module_3_Python_2/Pin_generator.py

def digit_to_char(d):
    # Mapping odd digits to their respective characters
    mapping = {1: 'a', 3: 'c', 5: 'e', 7: 'g', 9: 'i'}
    return mapping.get(d, str(d))

def single_digit_sum(n):
    # Summing the digits of a number until a single digit is obtained
    while n >= 10:
        n = sum(int(digit) for digit in str(n))
    return n

def generate_pin(numbers):
    result = []
    for number in numbers:
        # Compute the single digit sum for each number
        single_digit = single_digit_sum(number)
        # Convert the single digit to the corresponding character if it's odd
        result.append(digit_to_char(single_digit))
    return ''.join(result)

def main():
    # Take input from the user
    user_input = input()
    numbers = [int(num) for num in user_input.split()]
    
    # Generate and print the PIN
    pin = generate_pin(numbers)
    print( pin)

# Execute the main function
if __name__ == "__main__":
    main()